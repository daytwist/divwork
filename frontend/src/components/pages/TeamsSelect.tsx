import { useEffect, useState, useContext } from "react";
import { Link, useNavigate } from "react-router-dom";
import {
  Button,
  FormControl,
  InputLabel,
  MenuItem,
  OutlinedInput,
  Select,
  SelectChangeEvent,
  Stack,
  Typography,
} from "@mui/material";
import Grid2 from "@mui/material/Unstable_Grid2/Grid2";
import { Team } from "../../types/teamTypes";
import { BackButton } from "../ui/BackButton";
import { useFetchTeams } from "../../hooks/team/useFetchTeams";
import { LoadingColorRing } from "../ui/LoadingColorRing";
import { SnackbarContext } from "../../providers/SnackbarProvider";

export const TeamsSelect = () => {
  const { handleSetSnackbar } = useContext(SnackbarContext);
  const navigate = useNavigate();
  const [teamsData, isLoading] = useFetchTeams();

  const [teams, setTeams] = useState<Team[]>([]);
  const [teamId, setTeamId] = useState<string | number>("");
  const [teamName, setTeamName] = useState("");

  const ITEM_HEIGHT = 48;
  const ITEM_PADDING_TOP = 8;
  const MenuProps = {
    PaperProps: {
      style: {
        maxHeight: ITEM_HEIGHT * 4 + ITEM_PADDING_TOP,
        width: 250,
      },
    },
  };

  const checkIsSelectedTeam = () => {
    if (teamId) {
      navigate("/sign_up", {
        state: { teamId, teamName, isAdmin: false },
        replace: false,
      });
    } else {
      handleSetSnackbar({
        open: true,
        type: "error",
        message: "チームを選択してください",
      });
    }
  };

  const handleSelectChange = (event: SelectChangeEvent<string | number>) => {
    const { value } = event.target;
    setTeamId(value);

    const selectTeam: Team | undefined = teams.find((v) => v.id === value);
    console.log(selectTeam);

    if (selectTeam) {
      setTeamName(selectTeam.name);
    }
  };

  useEffect(() => {
    if (teamsData?.teams) {
      setTeams(teamsData.teams);
    }
  }, [teamsData?.teams]);

  if (isLoading) {
    return <LoadingColorRing />;
  }

  return (
    <Grid2 container direction="column" rowSpacing={3}>
      <Grid2 xs={12}>
        <Typography variant="h4" component="div" gutterBottom>
          所属チームの選択
        </Typography>
        <Typography variant="body1" component="div" gutterBottom>
          選択したチームでユーザー登録します。
        </Typography>
        <Typography
          variant="body2"
          color="text.secondary"
          component="div"
          gutterBottom
        >
          上限人数に達していないチームに参加することが出来ます。
        </Typography>
      </Grid2>
      <Grid2 xs={12}>
        <FormControl required sx={{ width: 300 }} color="secondary">
          <InputLabel id="team_label">チーム</InputLabel>
          <Select
            labelId="team_label"
            value={teamId}
            onChange={handleSelectChange}
            input={<OutlinedInput label="チーム" />}
            MenuProps={MenuProps}
          >
            {teams?.map((menu) => (
              <MenuItem key={menu.id} value={menu.id}>
                {menu.name}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </Grid2>
      <Grid2 xs={12}>
        <Stack direction="row" spacing={1}>
          <Button variant="contained" onClick={checkIsSelectedTeam}>
            次へ
          </Button>
          <BackButton />
        </Stack>
      </Grid2>
      <Grid2 xs={12}>
        <Typography variant="body1" component="div">
          または
        </Typography>
      </Grid2>
      <Grid2 xs={12}>
        <Button
          variant="contained"
          color="secondary"
          component={Link}
          to="/sign_up/teams/new"
        >
          新規チーム作成
        </Button>
      </Grid2>
    </Grid2>
  );
};
