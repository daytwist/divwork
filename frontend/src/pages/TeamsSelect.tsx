import { ChangeEvent, FC, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import {
  Button,
  Container,
  Grid,
  MenuItem,
  TextField,
  Typography,
} from "@mui/material";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../utils/axios";
import { Team, TeamsSelectResponse } from "../types";

const TeamsSelect: FC = () => {
  const [teams, setTeams] = useState<Team[]>([]);
  const [teamId, setTeamId] = useState<string>("");
  const [teamName, setTeamName] = useState<string>("");

  const options: AxiosRequestConfig = {
    url: "/teams/select",
    method: "GET",
  };

  useEffect(() => {
    axiosInstance(options)
      .then((res: AxiosResponse<TeamsSelectResponse>) => {
        console.log(res);
        setTeams(res.data.teams);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target;
    setTeamId(value);

    const teamIdNumber = Number(value);
    const selectTeam: Team | undefined = teams.find(
      (v) => v.id === teamIdNumber
    );
    console.log(selectTeam);

    if (selectTeam) {
      setTeamName(selectTeam.name);
    }
  };

  return (
    <Container maxWidth="sm">
      <Grid container direction="column" spacing={3}>
        <Grid item>
          <Typography variant="h4" component="div" gutterBottom>
            所属チームの選択
          </Typography>
          <Typography variant="subtitle1" component="div">
            選択したチームのユーザーを作成します。
          </Typography>
        </Grid>
        <Grid item>
          <TextField
            select
            label="チーム"
            sx={{ width: "25ch" }}
            name="id"
            value={teamId}
            defaultValue=""
            onChange={handleChange}
          >
            {teams?.map((menu) => (
              <MenuItem key={menu.id} value={menu.id}>
                {menu.name}
              </MenuItem>
            ))}
          </TextField>
        </Grid>
        <Grid item>
          <Link
            style={{ textDecoration: "none" }}
            to="/sign_up"
            key={teamId}
            state={{ teamId, teamName }}
          >
            <Button variant="contained" type="button">
              次へ
            </Button>
          </Link>
        </Grid>
      </Grid>
    </Container>
  );
};

export default TeamsSelect;
