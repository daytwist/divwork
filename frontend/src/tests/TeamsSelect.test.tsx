// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import TeamsSelect from "../pages/TeamsSelect";

describe("TeamsSelect", () => {
  test("スナップショット", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <TeamsSelect />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("チーム一覧の表示", async () => {
    render(<TeamsSelect />, { wrapper: BrowserRouter });
    const teamInput = screen.getByLabelText("チーム");

    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      userEvent.click(teamInput);
    });

    expect(await screen.findByText("TEAM_1")).toBeInTheDocument();
    expect(await screen.findByText("TEAM_2")).toBeInTheDocument();
  });
});
