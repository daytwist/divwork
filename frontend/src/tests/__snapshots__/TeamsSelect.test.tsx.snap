// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamsSelect スナップショット 1`] = `
<div
  className="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column css-s21l6u-MuiGrid2-root"
>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <div
      className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
    >
      所属チームの選択
    </div>
    <div
      className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
    >
      選択したチームでユーザー登録します。
    </div>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-ygea6l-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r0:"
        id=":r0:-label"
      >
        チーム
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
        variant="outlined"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r0:-label :r0:"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </div>
        <input
          aria-hidden={true}
          autoFocus={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          name="id"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              チーム
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <a
        href="/sign_up"
        onClick={[Function]}
        style={
          Object {
            "color": "black",
            "textDecoration": "none",
          }
        }
      >
        次へ
      </a>
    </button>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <div
      className="MuiTypography-root MuiTypography-subtitle1 css-iehxsa-MuiTypography-root"
    >
      または
    </div>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <div
      className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
    >
      新規チーム作成
    </div>
    <div
      className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
    >
      新しいチームの管理者としてユーザー登録します。
    </div>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-ygea6l-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r1:"
        id=":r1:-label"
      >
        新規チーム名
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id=":r1:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              新規チーム名
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid2-root MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      作成
    </button>
  </div>
</div>
`;
